### ISSUE TITLE
acceptance: OOM in allocator tests

### ISSUE THREAD

**Issue by cuongdo** [cockroachdb/cockroach#8126]
(Probably related to #8056 but not 100% sure so logging this as a separate issue.)

`TestUpreplicate_Small1To3` is now routinely failing due to out-of-memory errors. For example:

https://teamcity.cockroachdb.com/viewLog.html?buildId=6052&buildTypeId=Cockroach_AllocatorTest1to310g&tab=buildLog

I ran this test with an instrumented binary. The diffs are here:

https://gist.github.com/cuongdo/1280051280562b9a0aab21517a9c9be3

```
$ grep -c "creating preemptive"  cockroach.stderr
245
$ grep -c "creating snapshot" cockroach.stderr
455
```

So, 245 of 455 snapshots are preemptive. Despite this test going from 1 to 3 nodes (i.e. requiring 2 snapshots per range), the node is sometimes creating 4-5 snapshots for a given range (format is `count rangeID`):

```
$ grep "creating snapshot" cockroach.stderr  | awk '{print $10}' | sort | uniq -c  | sort -nrk 1 | head -20
   5 288
   5 251
   5 233
   4 97
   4 9
   4 84
   4 8
   4 74
   4 73
   4 70
   4 7
   4 56
   4 55
   4 52
   4 51
   4 47
   4 41
   4 40
   4 39
   4 38
```

Is there any other logging I should add to narrow down the issue? @tamird @petermattis 

Full logs here for the first node to crash:

[cockroach.stderr.txt](https://github.com/cockroachdb/cockroach/files/391039/cockroach.stderr.txt)

**Comment by petermattis** [comment]
I've mentioned this multiple times: the asynchronous sending of preemptive snapshots is not correct. After firing off the preemptive snapshot we immediately execute the Raft conf change to add the new replica. Raft will then race with the preemptive snapshot and often generate its own snapshot to bring the new replica up to date.

**Comment by petermattis** [comment]
Just re-read my comment and it was a tad aggressive. The upshot of your analysis is that it is close to my expectation and should be fixed by synchronously sending the preemptive snapshot.  Let's get this nailed next week.

**Comment by bdarnell** [comment]
The preemptive snapshot can be sent in parallel with the raft-generated snapshot for the same range, but that's only a factor of two. The bigger issue with async sending of preemptive snapshots is that this immediately frees up the replication queue to generate another async snapshot on a different range, and we can pile up many snapshots at once in this way. Either way, the right answer is to block on sending the snapshot, but cruder rate limiting in the replication queue might be an expedient interim solution.

**Comment by cuongdo** [comment]
This seems to not be happening after #8125
